src/Data/List/Ext/Properties.agda

```diff
@@ -54,11 +249,324 @@ module _ {a} {A : Set a} ⦃ _ : DecEq A ⦄ where
           helper : ∀ {l l' a} → a ∈ l ++ l' ⇔ (a ∈ l ⊎ a ∈ l')
           helper = mk⇔ (∈-++⁻ _) Data.Sum.[ ∈-++⁺ˡ , ∈-++⁺ʳ _ ]
 
--- TODO: stdlib?
-AllPairs⇒≡∨R∨Rᵒᵖ : ∀ {ℓ ℓ'} {A : Set ℓ} {R : A → A → Set ℓ'} {a b l}
-                 → AllPairs R l → a ∈ˡ l → b ∈ˡ l → a ≡ b ⊎ R a b ⊎ R b a
-AllPairs⇒≡∨R∨Rᵒᵖ [] = λ ()
-AllPairs⇒≡∨R∨Rᵒᵖ (x ∷ h) (here refl) (here refl) = inj₁ refl
-AllPairs⇒≡∨R∨Rᵒᵖ (x ∷ h) (here refl) (there b∈l) = inj₂ (inj₁ (All.lookup x b∈l))
-AllPairs⇒≡∨R∨Rᵒᵖ (x ∷ h) (there a∈l) (here refl) = inj₂ (inj₂ (All.lookup x a∈l))
-AllPairs⇒≡∨R∨Rᵒᵖ (x ∷ h) (there a∈l) (there b∈l) = AllPairs⇒≡∨R∨Rᵒᵖ h a∈l b∈l
+
+module _ {a} {A : Set a}  where
+
+  concatMap⁺ : {ys xs : List A} {f : A → List A} → xs ⊆ ys → concatMap f xs ⊆ concatMap f ys
+  concatMap⁺ = P.concat⁺ ∘ (P.map⁺ _)
+
+------------------------------------
+------- properties of insert -------
+------------------------------------
+module _ {a} {A : Set a} where
+  ∈→∷∈inserts : {ls : List (List A)} {x : A} → ∀ l → l ∈ ls → x ∷ l ∈ concatMap (insert x) ls
+  ∈→∷∈inserts {[] ∷ _} .[] (here refl) = here refl
+  ∈→∷∈inserts {(y ∷ _) ∷ _} .(y ∷ _) (here refl) = here refl
+  ∈→∷∈inserts {_ ∷ _} l (there l∈ls) = xs⊆ys++xs _ _ (∈→∷∈inserts l l∈ls)
+
+  ∈Allinsert : {l : List A} {x : A} → All (x ∈_) ((insert x) l)
+  ∈Allinsert {[]} = here refl All.∷ All.[]
+  ∈Allinsert {_ ∷ _} = (here refl) All.∷ ∈All→∈Allmap∷ ∈Allinsert
+
+  ∈Allinserts : {ls : List (List A)} {x : A} → All (x ∈_) (concatMap (insert x) ls)
+  ∈Allinserts {[]} = All.[]
+  ∈Allinserts {_ ∷ ls} = All++intro ∈Allinsert (∈Allinserts{ls})
+
+  ∈→∈Allinsert : {ys : List A} {y x : A} → x ∈ ys → All (x ∈_) (insert y ys)
+  ∈→∈Allinsert (here refl) = (there (here refl)) All.∷ ∈Allmap∷
+  ∈→∈Allinsert (there x∈ys) = (there (there x∈ys)) All.∷ (∈All→∈Allmap∷ (∈→∈Allinsert x∈ys))
+
+  ∈All→∈Allinserts : {ls : List (List A)} {y x : A} → All (x ∈_) ls → All (x ∈_) (concatMap (insert y) ls)
+  ∈All→∈Allinserts {.[]} All.[] = All.[]
+  ∈All→∈Allinserts {.(_ ∷ _)} (px All.∷ p) = All++intro (∈→∈Allinsert px) (∈All→∈Allinserts p)
+
+  Allinsert⊆∷ : {xs : List A} {x : A} → All(_⊆ (x ∷ xs)) (insert x xs)
+  Allinsert⊆∷ {[]} = id All.∷ All.[]
+  Allinsert⊆∷ {_ ∷ xs} = id All.∷ (All⊆-swap xs (All⊆→Allmap∷⊆∷ Allinsert⊆∷))
+
+  insert-decomp : {ls : List (List A)} {x : A} → ∀ ys → ys ∈ (concatMap (insert x) ls)
```

Comment:
I think there are at least three `insert-decomp` functions per direction, some with `concatMap` and some without. Please just take one for each direction without `concatMap`, and use these to prove what you need, everything will be way simpler.